src/esp/geo/VoxelUtils.cpp

```diff
@@ -0,0 +1,451 @@
+          closestCellGrid[i][j][k] = i_ahead;
+          sdfGrid[i][j][k] = intOrExtSign * i_dist;
+        } else if (j_dist <= i_dist && j_dist <= k_dist && j_dist <= cur_dist) {
+          closestCellGrid[i][j][k] = j_ahead;
+          sdfGrid[i][j][k] = intOrExtSign * j_dist;
+        } else if (k_dist <= i_dist && k_dist <= j_dist && k_dist <= cur_dist) {
+          closestCellGrid[i][j][k] = k_ahead;
+          sdfGrid[i][j][k] = intOrExtSign * k_dist;
+        } else {
+          sdfGrid[i][j][k] = intOrExtSign * cur_dist;
+        }
+      }
+    }
+  }
+  !Mn::Debug();
+}
+
+void generateDistanceGradientField(
+    std::shared_ptr<esp::geo::VoxelWrapper>& voxelWrapper,
+    const std::string& gridName) {
+  auto v_grid = voxelWrapper->getVoxelGrid();
+  auto m_voxelGridDimensions = v_grid->getVoxelGridDimensions();
+
+  // generate the ESDF if not already created
+  if (!v_grid->gridExists("ClosestBoundaryCell")) {
+    generateEuclideanDistanceSDF(voxelWrapper, "EuclideanSDF");
+  }
+
+  v_grid->addGrid<Mn::Vector3>(gridName);
+  auto gradientGrid = v_grid->getGrid<Mn::Vector3>(gridName);
+  auto closestBoundaryCell =
+      v_grid->getGrid<Mn::Vector3>("ClosestBoundaryCell");
+  std::vector<Mn::Vector3i> neighbors{
+      Mn::Vector3i(1, 0, 0),  Mn::Vector3i(-1, 0, 0), Mn::Vector3i(0, 1, 0),
+      Mn::Vector3i(0, -1, 0), Mn::Vector3i(0, 1, 0),  Mn::Vector3i(0, -1, 0)};
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        Mn::Vector3i index = Mn::Vector3i(i, j, k);
+        Mn::Vector3 result(0, 0, 0);
+        for (auto neighbor : neighbors) {
+          if (v_grid->isValidIndex(neighbor + index)) {
+            result += (Mn::Vector3(index + neighbor) -
+                       closestBoundaryCell[i + neighbor[0]][j + neighbor[1]]
+                                          [k + neighbor[2]]);
+          }
+        }
+
+        gradientGrid[i][j][k] =
+            result == Mn::Vector3() ? Mn::Vector3() : result.normalized();
+      }
+    }
+  }
+}
+
+std::vector<Mn::Vector3i> getVoxelSetFromBoolGrid(
+    std::shared_ptr<esp::geo::VoxelWrapper>& voxelWrapper,
+    const std::string& boolGridName) {
+  std::vector<Mn::Vector3i> voxelSet = std::vector<Mn::Vector3i>();
+  auto v_grid = voxelWrapper->getVoxelGrid();
+  auto m_voxelGridDimensions = v_grid->getVoxelGridDimensions();
+
+  assert(v_grid->gridExists(boolGridName));
+  auto boolGrid = v_grid->getGrid<bool>(boolGridName);
+  for (int i = 0; i < m_voxelGridDimensions[0]; i++) {
+    for (int j = 0; j < m_voxelGridDimensions[1]; j++) {
+      for (int k = 0; k < m_voxelGridDimensions[2]; k++) {
+        if (boolGrid[i][j][k]) {
+          voxelSet.push_back(Mn::Vector3i(i, j, k));
+        }
+      }
+    }
+  }
+  return voxelSet;
+}
+
+std::vector<Mn::Vector3i> getVoxelSetFromIntGrid(
```

Comment:
Looks like this could be templated to merge with float variant. 

Additionally a templated comparator could merge all 3. Python bindings could supply explicitly named variants with comparators so you don't need to bind the template or comparator functions.